### ISSUE TITLE
Add a podspec parameter to suppress all implicit linking

### ISSUE THREAD

**Issue by TGOS** [CocoaPods/CocoaPods#4494]
This will not break compatibility with existing pods as they won't have this setting. If set, don't link against anything not specified in `s.frameworks`. So if someone adds that setting and doesn't specify Cocoa or UIKit explicitly, the pod is not liked against them. If an umbrella header is generated, it also won't import `Cocoa.h` or `UIKit.h` unless these frameworks are listed in `s.frameworks`.

**Comment by neonichu** [comment]
I'd be weary to add a new DSL attribute for something that is ultimately very much an edge case.

**Comment by orta** [comment]
Same here

**Comment by TGOS** [comment]
Linking against Cocoa (which means liking against AppKit) has implicit consequences and you don't want to do that if you if you develop a UI-less launchd Agent, a system launchd Daemon (that will run with root privileges), a UI-less XPC service for your Application or just a simple command line tool. You may at most want to link against Foundation if you write it in Obj-C and sometimes you want to write it just in C (I've written XPC services in plain C several times) and then you may want just CoreFoundation or not even that. We have plenty of libraries that don't use Obj-C at all (they can be linked against Obj-C, but the can also be linked against just plain C or C++ code). But right now Cocoapods forces us to always link against Cocoa and always include Cocoa.h and this pretty much breaks everything, despite the fact that every security expert will rip off your head if you link a tool that will run with root privileges (e.g. a privileged helper tool, several of our apps need that) against Cocoa. 

So basically you are saying here, you are unwilling to make Cocoapods help development and dependency tracking for anything but the most stupid, primitive, one component UI only app? That's really sad! Considering all the other flaws and problems, I will recommend to move away of Cocoapods then and look for a more universal usable system.

**Comment by samdmarshall** [comment]
What you are seeing is automatic behavior from the compiler to resolved default libraries to link against a target. This is a behavior of Xcode and the compiler, not Cocoapods.

If you want to disable the automatic resolution of which libraries get linked, then you need to specify `LINK_WITH_STANDARD_LIBRARIES = NO` in your build settings or otherwise pass `-nostdlib` to the compiler when building. This means that ANY library you need will need to be specified in your "Link Binary With Libraries" build phase or otherwise specified as linker flags when building.

**Comment by TGOS** [comment]
Linking against Cocoa is not standard behavior. I'm referring to issue #4481 here. Standard library is `libSystem.dylib` and clang system libraries. The umbrella header that Cocoapods generates imports `<Cocoa/Cocoa.h>` (and Cocoapods generates it, not Xcode, not the system) and I was told that it does so because Cocoapods always links against Cocoa, regardless what you specify in `s.frameworks`, but this happens because Cocoapods does that explicitely. Xcode will not link against Cocoa, not unless it is told so either importing the Cocoa headers or by explicitly adding it as a dependent framework. If you do neither, your code is not linked against Cocoa.

**Comment by samdmarshall** [comment]
As I mentioned: the compiler will automatically add linker flags based on imports when the build setting `LINK_WITH_STANDARD_LIBRARIES` is enabled. If you also have `CLANG_MODULES_AUTOLINK` enabled then the system frameworks are added to the dependency info that is passed to the linker.

When both of these settings are enabled you will have a the system level libraries plus the system frameworks being passed to the linker via `-dependency-info`. Because `LINK_WITH_STANDARD_LIBRARIES` it will prioritize loading the libraries that are included for linking. Disabling `LINK_WITH_STANDARD_LIBRARIES` will remove that prioritization, however if you keep `CLANG_MODULES_AUTOLINK` enable, it is still passing the system frameworks to the linker. Disabling both of these should yield the build error of not being able to resolve `#import <Cocoa/Cocoa.h>`. Disabling `CLANG_MODULES_AUTOLINK` while `LINK_WITH_STANDARD_LIBRARIES` will still cause the compiler to resolve the necessary libraries for the linker to use. (you can verify this for yourself by passing `-t` to the linker via `OTHER_LDFLAGS`)

My understanding of this issue is that you do not want Cocoa.framework to be linked to your code. The problem is that it is being linked because of an import statement in your umbrella header. The reason why it is getting implicitly linked is because of the two build settings I mentioned (both of which are enabled in Xcode's default configuration. I would like to make the distinction that there is a different between the header generated by CocoaPods having `#import <Cocoa/Cocoa.h>` in it and CocoaPods adding a linker flag to explicitly link against that library. From what you have presented so far in this issue I haven't seen any evidence of CocoaPods performing any explicit linking. Any implicit linking behavior is the result of the build system (Xcode).

To be honest I do not know the reason why linking against Cocoa or UIKit is a requirement, beyond the fact this is explicitly a dependency manager for Cocoa. The use of the vast majority of the packages it provides are for applications or libraries that already use Cocoa, there is no additional cost incurred by asking the library to be linked and loaded at runtime because dyld won't load the same library twice.

From what you described the solution to this seems to be making the include in the umbrella header be optional if desired or use some means to strip that line from the generated umbrella header file.

**Comment by TGOS** [comment]
If Cocoapods is not explicitly linking Cocoa, then Samuel is wrong in #4481 when saying "because those frameworks have always been linked, regardless of whether they have been explicitly specified" as Cocoa has never been implicitly linked by Xcode and it still isn't. In very early versions of Xcode  it was linked because Apple put `-framework Cocoa` in the other linker flags for every new project. Later on Cocoa was just added to the "Link with Frameworks" build phase for every new project. And now it is usually linked because your code does `#import <Cocoa/Cocoa.h>` or `@import Cocoa;` and this causes the framework to automatically be liked.  But **none of that is implicit**, all of that happens explicitly. I named this issue *suppress all implicit linking* because of of the comments at #4481, which made me believe that Cocoapods is and always has been implicitly linking against Cocoa (and if that was the case, I was just looking for a totally backward compatible way to suppress that).

If you say that this linking is really only the result of `#import <Cocoa/Cocoa.h>` in the umbrella header (which was my problem to begin with), then we close this issue and I go back to #4481 to get rid of that import. This import is not required. CoreFoundation also has a `module.map` file and it has an umbrella header; take a look at it. Does it import Cocoa? Nope. Neither module support nor having an umbrella header make such an import necessary. Just removing it might break code that incorrectly relied upon that import, I'm totally with the developers here. However, that import is not even documented anywhere. This import is implicit and while implicit behavior itself isn't good to begin with, it's horrible when undocumented. So I wonder why it isn't possible to just suppress it. This issue was never really about the linking, it was actually about the import that (as you say) causes the linking.

The main problem with that import is, that when I import any of the framework headers into a `.c` file (plain C), then the umbrella header is imported instead. This is now expected behavior. If a framework has a module.map and also an umbrella header, then importing any header that is also imported by the umbrella header will import the umbrella header itself. This is usually nice and desired, but importing `Cocoa.h` into a plain C file will obviously not work as clang will not support any Obj-C constructs when you build a C file. I can protect all my Obj-C headers against such a fatal import (`#ifdef __OBJC__`) but that won't help if `Cocoa.h` is always imported, no matter what I do. There's a reason why Apple puts `#ifdef __OBJC__` into PCH files, so you can separated Obj-C imports from other imports (as the PCH file will also apply to C and C++ files).

**Comment by samdmarshall** [comment]
I think there are some assumptions about wording/terminology here. My understanding of explicit linking is that a flag is passed (such as `-framework Cocoa` as you mentioned or by having Xcode do it by adding to the linked library list); and implicit linking is when no flag is passed but it will be linked anyway as a requirement of some other part of the process. Since (as least as I understand it) there are no linker flags that call to have Cocoa.framework linked, this is indeed *implicit* linking behavior. Since the build system is responsible for this (as mentioned: the fact of the import statement being there is not causing this, the build settings are), CocoaPods cannot control how your app's target is configured. So while the CocoaPods write the import into the umbrella header, the build system should ultimately make the choice as to how linking is performed. Based on that, it seems that the implied behavior was to have Cocoa get linked regardless as a matter of convenience to the library developer.

You are right, the importing of the Cocoa framework header into your C code is not desirable; and that seems to be the issue that should be addressed here. What I have mentioned so far is my understanding of the problem based on your description. I think we may need some confirmation that removing that line from your umbrella header does, in-fact, resolve the extra linking behavior you are seeing before proceeding to a resolution based on this information. If you are able to test this out or provide us with a sample project that reproduces this behavior that would be most helpful. If this does resolve your issue then I am going to defer to members of CocoaPods Core team as to how they would like to address this. 

Thanks for your patience on this!

**Comment by TGOS** [comment]
Regarding implicit: To me implicit means something happens even though I have not done anything to make it happen. Implicit is that clang will always link libSystem unless I tell it explicitly not to do that (which it will always do, even if not a single libSystem header is included in my code and no flag is passed). Importing a headers is something I do explicitly, that means, if it was me importing it myself. The linking then happens automatically, but not truly implicit (it is caused by my explicit import and automatic is no synonym for implicit).

Okay, the issues is not really about linking. As you pointed out, thin linking is only a result of the import. And the import on it's own is not a problem but it makes it impossible to have a framework that can also be used by C code. I created a demo project to demonstrate that problem. The targets XPC_A and XPC_B are both linked with the pod but XPC_A fails to be built because it contains a C file and even thought the C file never directly imports any Obj-C H file, the way how CP sets up the umbrella header causes compilation to break. As a counter sample there is a target XPC_C with a similar setup, except that I made the framework myself (it's not a from a pod) and can also be linked to XPC_D and XPC_E (so it can be used in all places the pod can be used as well) - also it follows Apple's recommendation how to write umbrella headers, while CP violates it (Apple says import all public H files as `#import <FrameworkName/Public.h>` yet CP imports them as `#import "Public.h"`).

I found out that I can work around the problem with my own map file and hand written umbrella header, but basically CP breaks C code compatibility whenever it writes own Umbrella headers, even if the final project would link against Cocoa, it won't eve build and get to the linking stage.
